infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesDeployments.java

```diff
@@ -125,26 +125,43 @@ public Pod deploy(Pod pod) throws InfrastructureException {
     ObjectMeta metadata = pod.getMetadata();
     PodSpec podSpec = pod.getSpec();
     podSpec.setRestartPolicy("Always"); // Only allowable value
+    final CompletableFuture<Pod> createFuture = new CompletableFuture<>();
+    final Watch createWatch =
+        clientFactory
+            .create(workspaceId)
+            .pods()
+            .inNamespace(namespace)
+            .watch(new CreateWatcher(createFuture, workspaceId));
     try {
-      Deployment deployment =
-          clientFactory
-              .create(workspaceId)
-              .extensions()
-              .deployments()
-              .inNamespace(namespace)
-              .createNew()
-              .withMetadata(metadata)
-              .withNewSpec()
-              .withReplicas(1)
-              .withNewTemplate()
-              .withMetadata(metadata)
-              .withSpec(podSpec)
-              .endTemplate()
-              .endSpec()
-              .done();
-      return awaitAndGetPod(deployment);
+      clientFactory
+          .create(workspaceId)
+          .extensions()
+          .deployments()
+          .inNamespace(namespace)
+          .createNew()
+          .withMetadata(metadata)
+          .withNewSpec()
+          .withReplicas(1)
+          .withNewTemplate()
+          .withMetadata(metadata)
+          .withSpec(podSpec)
+          .endTemplate()
+          .endSpec()
+          .done();
+      return createFuture.get(POD_CREATION_TIMEOUT_MIN, TimeUnit.MINUTES);
     } catch (KubernetesClientException e) {
       throw new KubernetesInfrastructureException(e);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new InfrastructureException(
+          "Interrupted while waiting for Pod creation. " + e.getMessage());
+    } catch (ExecutionException e) {
+      throw new InfrastructureException(
+          "Error occured while waiting for Pod creation. " + e.getMessage());
```

Comment:
Does `e.getMessage` contain a meaningful message or it is needed to fetch the message of the cause exception? Like in [KubernetesNamespace](https://github.com/eclipse/che/blob/3ad17d60c9b5c5b09ab508296f1e0b15ff054d06/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java#L251). Please check which way is better.